### ISSUE TITLE
`<<` operator doesn't work well with `AbstractInt` changes in 0.19

### ISSUE THREAD

**Issue by SludgePhD** [gfx-rs/wgpu#5085]
**Description**
Some of the changes to support `AbstractInt` seem to have regressed which shader code is accepted by naga, requiring fairly ugly workarounds.

**Repro steps**

This function used to be accepted fine on wgpu 0.18:

```wgsl
fn huff_extend(v: i32, t: u32) -> i32 {
    let vt = 1 << (t - 1u);
    return select(v, v + (-1 << t) + 1, v < vt);
}
```

But isn't anymore on 0.19, yielding the following errors:

```
Shader 'huffman' parsing error: failed to convert expression to a concrete type: Cannot cast scalar components of expression `[7] Binary { op: ShiftLeft, left: [3], right: [6] }` to type `i32`
    ┌─ wgsl:247:14
    │
247 │     let vt = 1 << (t - 1);
    │              ^^^^^^^^^^^^ this expression has type {AbstractInt}
    │
    = note: the expression should have been converted to have i32 scalar type
```

and

```
Shader validation error: 
    ┌─ huffman:246:1
    │  
246 │ ╭ fn huff_extend(v: i32, t: u32) -> i32 {
247 │ │     let vt = i32(1) << (t - 1u);
248 │ │     return select(v, v + (-1 << t) + 1, v < i32(vt));
    │ │                           ^^ naga::Expression [7]
    │ ╰─────────────────────────────────────────────────────^ naga::Function [5]


    Function [5] 'huff_extend' is invalid
    Expression [7] is invalid
    Abstract types may only appear in constant expressions
```

Instead, the following changes are needed:

```wgsl
fn huff_extend(v: i32, t: u32) -> i32 {
    let vt = i32(1) << (t - 1u);
    return select(v, v + (i32(-1) << t) + 1, v < vt);
}
```

**Expected vs observed behavior**

This might be expected fallout from the abstract type changes, but I figured I would file this issue anyways just to make sure.

**Comment by rcoreilly** [comment]
Yeah this is very cumbersome to work around (and unclear as to what is needed to make the compiler happy).

**Comment by jamienicol** [comment]
I've been digging in to this a bit today, unfortunately haven't fully figured it out but can pick it back up in the new year.

First of all, let's look at a related issue, simpler than the one in the example above. This shader fails:
```wgsl
fn foo() {
    let x = 1 << 1u;
}
```

This is because we haven't implemented bitshifting for abstract ints in `ConstantEvaluator::binary_op`. That's a really easy fix.

Now the problem described above:
```wgsl
fn foo(a: u32) {
    let x = 1 << a;
}
```

gives the error:
```
Could not parse WGSL:
error: failed to convert expression to a concrete type: Cannot cast scalar components of expression `[2] Binary { op: ShiftLeft, left: [1], right: [0] }` to type `i32`
  ┌─ in.wgsl:2:13
  │
2 │     let x = 1 << a;
  │             ^^^^^^ this expression has type {AbstractInt}
  │
  = note: the expression should have been converted to have i32 scalar type
```

This error comes from [here](https://github.com/gfx-rs/wgpu/blob/b1ca9dfd1fd8b4fdd61ecf1e3dafbd93f389ec41/naga/src/front/wgsl/lower/conversion.rs#L262) after `ConstantEvaluator::cast()` returns an `ConstantEvaluatorError::InvalidCastArg`.

I was curious why this affects bit shifts but not other binops. I believe the reason is that for other binops, eg addition, either a) both operands are abstract, meaning we can constantly evaluate them to a literal, so `cast()` succeeds. Or b) at least one operand is concrete, meaning the other operand is concretized [here](https://github.com/gfx-rs/wgpu/blob/b1ca9dfd1fd8b4fdd61ecf1e3dafbd93f389ec41/naga/src/front/wgsl/lower/mod.rs#L2090) and the result of the operation is therefore concrete. Bitshifting is special in that if we shift an abstract value by a variable then the result is both abstract and *not* const-evaluable

**Comment by jimblandy** [comment]
If you look at the WGSL spec, the `<<` operator is really different from all the other binary operators, in that the right operand is always a shift distance, and only the left operand is relevant to the result type. So the handling of the other operators probably isn't a helpful model at all.

**Comment by jamienicol** [comment]
Yes precisely. Other binops can infer the result type from either arg individually, or if both are abstract can be const evaled out. Shifting is unique among operators in that regard afaict. Though I think builtin functions will be pretty similar to bitshifts

**Comment by jimblandy** [comment]
I think the problem is simply that `ConstantEvaluator::binary_op` doesn't cover bit shift operators with left operands that are AbstractInt.

**Comment by jamienicol** [comment]
> I think the problem is simply that `ConstantEvaluator::binary_op` doesn't cover bit shift operators with left operands that are AbstractInt.

That solves the issue when the right operand is a constant, but the example given at the top of this bug still fails as the right operand is not constant.

I'll create a PR for the right-operand-is-const case as it's trivial, but we should leave this bug open.
